pkg/controller/config/config_controller.go

```diff
@@ -249,50 +251,78 @@ func (r *ReconcileConfig) Reconcile(request reconcile.Request) (reconcile.Result
+	} else {
+		// The watch set _has_ changed, so recalculate the replay set.
+		r.needsReplay = nil
+		r.needsWipe = true
 	}
 
 	// --- Start watching the new set ---
 
 	// This must happen first - signals to the opa client in the sync controller
 	// to drop events from no-longer-watched resources that may be in its queue.
-	needReplay := r.watched.Union(newSyncOnly)
+	if r.needsReplay == nil {
+		r.needsReplay = r.watched.Intersection(newSyncOnly)
+	}
 	r.watched.Replace(newSyncOnly)
 
 	// swapping with the new excluder
 	r.processExcluder.Replace(newExcluder)
 
 	// *Note the following steps are not transactional with respect to admission control*
 
-	// Wipe all data to avoid stale state
-	if _, err := r.opa.RemoveData(context.Background(), target.WipeData{}); err != nil {
-		return reconcile.Result{}, err
+	// Wipe all data to avoid stale state if needed. Happens once per watch-set-change.
+	if err := r.wipeCacheIfNeeded(ctx); err != nil {
+		return reconcile.Result{}, fmt.Errorf("wiping opa data cache: %w", err)
 	}
 
-	// reset sync cache before sending the metric
-	r.syncMetricsCache.ResetCache()
-	r.syncMetricsCache.ReportSync(&syncc.Reporter{Ctx: context.TODO()})
-
 	// Important: dynamic watches update must happen *after* updating our watchSet.
 	// Otherwise the sync controller will drop events for the newly watched kinds.
+	// Defer error handling so object re-sync happens even if the watch is hard
+	// errored due to a missing GVK in the watch set.
 	if err := r.watcher.ReplaceWatch(newSyncOnly.Items()); err != nil {
 		return reconcile.Result{}, err
 	}
 
 	// Replay cached data for any resources that were previously watched and still in the watch set.
-	// This is necessary because we wiped their data from Opa above.
+	// This is necessary because we wipe their data from Opa above.
 	// TODO(OREN): Improve later by selectively removing subtrees of data instead of a full wipe.
-	if err := r.replayData(context.TODO(), needReplay); err != nil {
+	if err := r.replayData(ctx); err != nil {
 		return reconcile.Result{}, fmt.Errorf("replaying data: %w", err)
 	}
 
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileConfig) wipeCacheIfNeeded(ctx context.Context) error {
+	if r.needsWipe {
+		if _, err := r.opa.RemoveData(ctx, target.WipeData{}); err != nil {
+			return err
+		}
+
+		// reset sync cache before sending the metric
+		r.syncMetricsCache.ResetCache()
+		r.syncMetricsCache.ReportSync(&syncc.Reporter{Ctx: context.TODO()})
+
+		r.needsWipe = false
+	}
+	return nil
+}
+
 // replayData replays all watched and cached data into Opa following a config set change.
 // In the future we can rework this to avoid the full opa data cache wipe.
-func (r *ReconcileConfig) replayData(ctx context.Context, w *watch.Set) error {
-	for _, gvk := range w.Items() {
+func (r *ReconcileConfig) replayData(ctx context.Context) error {
+	if r.needsReplay == nil {
+		return nil
+	}
+	for _, gvk := range r.needsReplay.Items() {
```

Comment:
What if a gvk doesnt exist in the cluster anymore but is in the config, then the replayData will continue to return an error and skip all other gvks?

Reply:
Correct, it's not ideal but it's better than the current behavior, where we are stuck entirely after 1 failure (at least there is some self-healing now).

This mitigates the problem, but the long term fix is to remove the "delete all" logic entirely and replace it with pruning, which brycecr@ is looking at designing.